src/openforms/authentication/views.py

```diff
@@ -339,6 +345,14 @@ def _handle_return(self, request: Request, slug: str, plugin_id: str):
         if hasattr(request, "session") and FORM_AUTH_SESSION_KEY in request.session:
             authentication_success.send(sender=self.__class__, request=request)
 
+        if initial_data_reference:
+            # Pass the initial data reference back to the SDK, to allow sending it
+            # with the submission create call
+            return HttpResponseRedirect(
+                furl(response.url)
+                .add({"initial_data_reference": initial_data_reference})
+                .url
+            )
```

Comment:
IMO this is too naive and we should probably look at a better/proper way to capture some state to re-use when redirecting back to the frontend.

If you look at `BasePlugin.handle_return`, the only guarantee is that it will return an `HttpResponse` - you have no guarantees about it being a redirect. Perhaps we must take a critical look at this, it's quite old code and formalize a couple of things.

I'm also a bit concerned about the amount of data we put in the session, it is sensitive to race conditions if multiple authentication requests start and concurrency on session data is quite bad itself.

Maybe an approach where the generic code prepares the return URL and passes it as a parameter to the plugin would be much more robust, that makes it properly stateless and removes the duplication of URL-encoding parameters that happens in plugins.

Ultimately I was surprised to see changes to the `authentication` module to handle this ticket - the responsibility of this module is authenticating users and it should not have any knowledge about the rest of the system. Giving the `initial_data_reference` special treatment doesn't fit the separation of concerns, and it's also not documented in the API documentation for these endpoints (nor should it, in my opinion).

The authentication start view gets the `form_url = request.GET.get("next")` parameter, which should contain *all* the information already (including GET parameters) to send the user back to the form after authenticating.

Reply:
To add to this - nowhere do we seem to add the `_start` parameter in the backend, this is passed in from the frontend code, so I would really expect the `initial_data_reference` to work the same way, it's also just a query string parameter :thinking:

Reply:
you're right, I added it to the wrong queryParams in the SDK (I should have added it to the query params of the nextUrl :facepalm:). Made a PR for it https://github.com/open-formulieren/open-forms-sdk/pull/727 and I'll remove the backend changes